src/java/org/apache/cassandra/service/snapshot/ClearSnapshotTask.java

```diff
@@ -0,0 +1,196 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service.snapshot;
+
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.config.DurationSpec;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.utils.Clock;
+
+public class ClearSnapshotTask extends AbstractSnapshotTask<Void>
+{
+    private static final Logger logger = LoggerFactory.getLogger(ClearSnapshotTask.class);
+
+    private final SnapshotManager manager;
+    private final Predicate<TableSnapshot> predicateForToBeCleanedSnapshots;
+    private final boolean deleteData;
+
+    public ClearSnapshotTask(SnapshotManager manager,
+                             Predicate<TableSnapshot> predicateForToBeCleanedSnapshots,
+                             boolean deleteData)
+    {
+        super(null);
+        this.manager = manager;
+        this.predicateForToBeCleanedSnapshots = predicateForToBeCleanedSnapshots;
+        this.deleteData = deleteData;
+    }
+
+    @Override
+    public SnapshotTaskType getTaskType()
+    {
+        return SnapshotTaskType.CLEAR;
+    }
+
+    @Override
+    public Void call()
+    {
+        Set<TableSnapshot> toRemove = new HashSet<>();
+
+        for (TableSnapshot snapshot : new GetSnapshotsTask(manager, predicateForToBeCleanedSnapshots, false).call())
+        {
+            logger.debug("Removing snapshot {}{}", snapshot, deleteData ? ", deleting data" : "");
+
+            toRemove.add(snapshot);
+
+            if (deleteData)
+            {
+                for (File snapshotDir : snapshot.getDirectories())
+                {
+                    try
+                    {
+                        removeSnapshotDirectory(snapshotDir);
+                    }
+                    catch (Throwable ex)
+                    {
+                        logger.warn("Unable to remove snapshot directory {}", snapshotDir, ex);
+                    }
+                }
+            }
+        }
+
+        manager.getSnapshots().removeAll(toRemove);
+
+        return null;
+    }
+
+    /**
+     * Returns predicate which will pass the test when arguments match.
+     *
+     * @param tag name of snapshot
+     * @param options options for filtering
+     * @param keyspaceNames names of keyspaces a snapshot is supposed to be from
+     * @return predicate which will pass the test when arguments match.
+     */
+    public static Predicate<TableSnapshot> getPredicateForCleanedSnapshots(String tag, Map<String, Object> options, String... keyspaceNames)
```

Comment:
do we need this as public? maybe make it package-private instead?

Reply:
made package protected